packages/@aws-cdk/aws-codepipeline-actions/test/codecommit/test.codecommit-source-action.ts

```diff
@@ -109,6 +109,64 @@ export = {
 
       test.done();
     },
+
+    'exposes variables for other actions to consume'(test: Test) {
+      const stack = new Stack();
+
+      const sourceOutput = new codepipeline.Artifact();
+      const codeCommitSourceVariables = new cpactions.CodeCommitSourceVariables();
+      new codepipeline.Pipeline(stack, 'Pipeline', {
+        stages: [
+          {
+            stageName: 'Source',
+            actions: [
+              new cpactions.CodeCommitSourceAction({
+                actionName: 'Source',
+                repository: new codecommit.Repository(stack, 'MyRepo', {
+                  repositoryName: 'my-repo',
+                }),
+                output: sourceOutput,
+                variables: codeCommitSourceVariables,
```

Comment:
Since this is a CodeCommitSourceAction, why isn't this just a `boolean` that indicates whether to expose  `CodeCommitSourceVariables`?

Something like:

```
const action = new cpactions.CodeCommitSourceAction({
  // ...
  variables: true
});
```

And then, `action.variables` is of type `CodeCommitSourceVariables`

Reply:
I didn't do it that way, because that would force you to assign each action to a local variable. We don't want to do that, because you often want to create a pipeline in one expression, for example:

```ts
        import codepipeline = require('@aws-cdk/aws-codepipeline');
        import cpactions = require('@aws-cdk/aws-codepipeline-actions');

        const sourceOutput = new codepipeline.Artifact('SourceOutput');
        new codepipeline.Pipeline(this, 'Pipeline', {
            stages: [
                {
                    stageName: 'Source',
                    actions: [
                        new cpactions.GitHubSourceAction({
                            // ...
                            actionName: 'GitHub_Source',
                            output: sourceOutput,
                        }),
                    ],
                },
                {
                    stageName: 'Build',
                    actions: [
                        new cpactions.CodeBuildAction({
                            // ...
                            actionName: 'Build_CodeBuild',
                            input: sourceOutput,
                        }),
                    ],
                },
            ],
        });
```

I modeled variables similarly to how we modeled artifacts. If you remember, artifacts used to be read-only properties of the actions, like you are proposing for variables. Which meant you had to do something like this instead:

```ts
        const sourceAction = new cpactions.GitHubSourceAction({
            // ...
            actionName: 'GitHub_Source',
        });
        new codepipeline.Pipeline(this, 'Pipeline', {
            stages: [
                {
                    stageName: 'Source',
                    actions: [
                        sourceAction,
                    ],
                },
                {
                    stageName: 'Build',
                    actions: [
                        new cpactions.CodeBuildAction({
                            // ...
                            actionName: 'Build_CodeBuild',
                            input: sourceAction.output,
                        }),
                    ],
                },
            ],
        });
```

However, the feedback we got (and you were one of the more vocal users giving this feedback 🙂) was that forcing this local variable creation was too much friction. And for that reason, we changed to having stand-alone artifacts in #2338.

That's why I modeled variables in a way that does _not_ force you to assign actions to intermediate variables.

Reply:
But then this API forces the use of a local variable for the `Variables` object, so what's the difference? If I already need a local variable to access variables, I would rather just use a local variable for the `Action` and access `action.variables`. Makes more sense to me then the current API.

Reply:
Because a local variable for one property is significantly less friction than forcing a local variable for the entire expression...?

Let's say you have this pipeline:

```ts
new codepipeline.Pipeline(this, 'Pipeline', {
  stages: [
    {
      stageName: 'Source',
      actions: [
        new cpactions.GitHubSourceAction({
          // ...
          actionName: 'GitHub_Source',
          output: sourceOutput,
        }),
      ],
    },
    {
      stageName: 'Build',
      actions: [
        new cpactions.CodeBuildAction({
          // ...
          actionName: 'Build_CodeBuild',
          input: sourceOutput,
        }),
      ],
    },
  ],
});
```

Now, if you want to use variables with the current design, you just have to do:

```ts
const gitHubSourceVariables = new cpactions.GitHubSourceVariables();
new codepipeline.Pipeline(this, 'Pipeline', {
  stages: [
    {
      stageName: 'Source',
      actions: [
        new cpactions.GitHubSourceAction({
          // ...
          actionName: 'GitHub_Source',
          output: sourceOutput,
          variables: gitHubSourceVariables,
        }),
      ],
    },
    {
      stageName: 'Build',
      actions: [
        new cpactions.CodeBuildAction({
          // ...
          actionName: 'Build_CodeBuild',
          input: sourceOutput,
          environmentVariables: { A: { value: gitHubSourceVariables.commitId } },
        }),
      ],
    },
  ],
});
```

I would argue it's a relatively frictionless, small change.

With your proposal, using variables would look like this:

```ts
const sourceAction = new cpactions.GitHubSourceAction({
  // ...
  actionName: 'GitHub_Source',
  output: sourceOutput,
  variables: true,
});
const gitHubSourceVariables = new cpactions.GitHubSourceVariables();
new codepipeline.Pipeline(this, 'Pipeline', {
  stages: [
    {
      stageName: 'Source',
      actions: [
        sourceAction
      ],
    },
    {
      stageName: 'Build',
      actions: [
        new cpactions.CodeBuildAction({
          // ...
          actionName: 'Build_CodeBuild',
          input: sourceOutput,
          environmentVariables: { A: { value: sourceAction.variables.commitId } },
        }),
      ],
    },
  ],
});
```

This forces a complete re-organization of the of the code. I would argue that's it causes a ton of friction.

Also, I don't understand why, if we want this API shape for variables, we would do something different for artifacts - it seems inconsistent... are you saying the changes in #2338 were a mistake?

Reply:
I can't say I love our artifacts API, and I also think think this is a different situation. With artifacts, there is a notion of "wiring" the artifacts from the output of one action to the input of the next. Here, there is no meaning to the value you pass as input, it's purely an output API (right?). Furthermore, I would argue that we can completely omit `variables` from `XxxActionProps` and instantiate the relevant variables objects lazily as soon as users access the action's `variables` property, the input property is completely redundant.

```ts
const action = new GitHubSourceAction({
  // ...
  actionName: 'GitHubSource',
  output: sourceOutput,
  // no need to specify "variables: true" or anything here!
});

// the `variables` getter will lazily render the variables and will always be defined
const commitId = action.variables.commitId; // this will "just work"
```

Reply:
> I can't say I love our artifacts API, and I also think think this is a different situation. With artifacts, there is a notion of "wiring" the artifacts from the output of one action to the input of the next. Here, there is no meaning to the value you pass as input, it's purely an output API (right?).

No, there is an input component. Each action that produces variables needs to have a (pipeline unique) namespace. It's an optional property when constructing the variables object:

```ts
// we generate a namespace name for you, based on the action and stage names:
const gitHubSourceVariables = new cpactions.GitHubSourceVariables();

// explicitly set namespace name:
const gitHubSourceVariables = new cpactions.GitHubSourceVariables('MyNamespace');
```

Reply:
I'd model it through an optional property on the action:

```ts
const sourceAction = new cpactions.GitHubSourceAction({
  // ...
  actionName: 'GitHub_Source',
  output: sourceOutput,
  variablesNamespace: 'MyNamespace',
});
```

Reply:
That's literally the same API as we used to have for artifacts before  #2338 was merged:

```ts
const source = new codepipeline_actions.GitHubSourceAction({
  // ....
  outputName: 'Output', // optional
});

new codepipeline_actions.CodeBuildAction({
  // ...
  input: source.outputArtifact,
});
```

Why would we go back to an API that we didn't like, that received negative feedback from our customers on usability (including from you), and that we made a big effort to change?

Reply:
There are many differences between these APIs:
1. You can't use CodePipelines without artifacts, but most users will probably never use variables. This means that the way artifacts are accessed would impact *all* users, as oppose to variables which are only relevant to a small fraction of the users.
2. Artifacts have both inputs and outputs, variables are only outputs.
3. Artifacts are wired across actions. This lends itself nicely to the concept of a "handle" that you pass in to enable the wiring. 

For the record: I don't like our current artifacts API, but I've made my peace with it. I still think there are ergonomic improvements we can do around the artifacts API, which we discussed in the past (like making some of these optional in case they can be deduced automatically).

Reply:
> 2. Artifacts have both inputs and outputs, variables are only outputs.

That's not true, though - an artifact can only ever be used as an output once. We even have validation that checks that. In that sense, the two are exactly the same.

> 3. Artifacts are wired across actions. This lends itself nicely to the concept of a "handle" that you pass in to enable the wiring.

But that's exactly what variables are used for as well. There is no point in using variables if they are not wired across actions - it would be pointless to use them in the first place.

> For the record: I don't like our current artifacts API, but I've made my peace with it. I still think there are ergonomic improvements we can do around the artifacts API, which we discussed in the past (like making some of these optional in case they can be deduced automatically).

I think automatic deduction is completely orthogonal to this discussion, though (by which I mean, we can implement automatic deduction for both of those API shapes).

------

Can you elaborate on how is the alternative of "variables as properties of action instances" superior to the proposal in this PR? I think I've made a pretty solid case on why I think this PR is the way to go (consistent with the artifacts API, does not force you to assign actions to intermediate variables, leverages the learnings from #2338 about what experience customers want), but I don't think I've seen a specific argument for "variables as properties of action instances" other than "Makes more sense to me then the current API".

Reply:
I think the proposed API is less intuitive and less ergonomic. 

For example, the proposed API will allow me to do this:

```ts
const sourceAction = new GitHubSourceAction({
  // ...
  actionName: 'source',
  output: sourceOutput,
});

const anotherAction = new CodeBuildAction({
  // ...
  actionName: 'build',
  input: sourceOutput,
  environmentVariables: { 
    COMMIT_ID: { value: sourceAction.variables?.commitId } 
  },
});
```

Which will silently assign `undefined` to the `COMMIT_ID` environment variable of the build action.

And then, with the correct usage of the proposed API, which looks like this:

```ts
const sourceActionVariables = new GitHubSourceVariables();
const sourceAction = new GitHubSourceAction({
  // ...
  actionName: 'source',
  output: sourceOutput,
  varaibles: sourceActionVariables
});

const anotherAction = new CodeBuildAction({
  // ...
  actionName: 'build',
  input: sourceOutput,
  environmentVariables: { 
    COMMIT_ID: { value: sourceAction.variables?.commitId } 
  },
});
```

I ends up in a situation where I have `sourceActionVariables`, but I am not actually using it for anything, so I am not clear why do I even need to instantiate it?? Feels odd.

----

The alternative in my mind is much more intuitive. For example, the above usage which broke in the first API will "just work" with the second API. 

```ts
const sourceAction = new GitHubSourceAction({
  // ...
  actionName: 'source',
  output: sourceOutput,
});

const anotherAction = new CodeBuildAction({
  // ...
  actionName: 'build,
  input: sourceOutput,
  environmentVariables: { 
    COMMIT_ID: { value: sourceAction.variables.commitId } 
  },
});
```

I just reference `variables.commitId` and the source action will automatically define the variables for me.

It's also less verbose and ceremonies, and includes less repetition.

Granted, the API I am proposing makes it impossible to reference variables without assigning the action to a variable, but there are other things in the action that can't be referenced if it is used declaratively. 

I believe that since variables are not required for every pipeline definition, it's a decent tradeoff, especially given we get a much cleaner and less error-prone developer experience.

Reply:
Your first example does not accurately show the API this PR proposes.

This is its correct usage:

```ts
const sourceActionVariables = new GitHubSourceVariables();
const sourceAction = new GitHubSourceAction({
  // ...
  actionName: 'source',
  output: sourceOutput,
  variables: sourceActionVariables,
});

const anotherAction = new CodeBuildAction({
  // ...
  actionName: 'build',
  input: sourceOutput,
  environmentVariables: { 
    COMMIT_ID: { value: sourceActionVariables.commitId },
  },
});
```

* There's no `sourceAction.variables?.commitId` (action classes don't get any additional properties in this PR)
* It's not possible for `sourceActionVariables.commitId` to return `undefined`
* You would get a validation error if you tried to use `sourceActionVariables.commitId` without passing `sourceActionVariables` in the `variables` property when creating an action
* It doesn't make any sense to instantiate a variables class like `GitHubSourceVariables` without referencing the instance later

Does this clarification change your stance in any way?

Reply:
> Granted, the API I am proposing makes it impossible to reference variables without assigning the action to a variable, but there are other things in the action that can't be referenced if it is used declaratively.

Can you give an example?